src/stream/src/executor/sort.rs

```diff
@@ -0,0 +1,506 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::ops::Bound;
+
+use futures::StreamExt;
+use futures_async_stream::try_stream;
+use risingwave_common::array::{Op, Row, StreamChunk};
+use risingwave_common::buffer::Bitmap;
+use risingwave_common::catalog::Schema;
+use risingwave_common::types::Datum;
+use risingwave_common::util::select_all;
+use risingwave_storage::table::streaming_table::state_table::StateTable;
+use risingwave_storage::StateStore;
+
+use super::error::StreamExecutorError;
+use super::{
+    expect_first_barrier, ActorContextRef, BoxedExecutor, BoxedMessageStream, Executor, Message,
+    PkIndices, Watermark,
+};
+
+/// [`SortBufferKey`] contains a record's timestamp and pk.
+type SortBufferKey = (Datum, Row);
+/// [`SortBufferValue`] contains a record's operation and value, and also a flag indicating whether
+/// the record has been persisted to storage.
+type SortBufferValue = (Op, Row, bool);
+
+/// [`SortExecutor`] consumes unordered input data and outputs ordered data to downstream.
+pub struct SortExecutor<S: StateStore> {
+    context: ActorContextRef,
+
+    /// We make it `Option` here due to lifetime restrictions. It should always be `Some`.
+    input: Option<BoxedExecutor>,
+
+    pk_indices: PkIndices,
+
+    identity: String,
+
+    schema: Schema,
+
+    state_table: StateTable<S>,
+
+    /// The index of the column on which the sort executor sorts data.
+    sort_column_index: usize,
+
+    /// Stores data in memory ordered by the column indexed by `sort_column_index`. Once a
+    /// watermark of `sort_column_index` arrives, data below watermark (i.e. value of that column
+    /// being less than or equal to the watermark) should be sent to downstream and cleared from
+    /// the buffer.
+    buffer: BTreeMap<SortBufferKey, SortBufferValue>,
```

Comment:
There's an exhausting trade-off for which structure to use for the in-memory cache. 🥵 For example, up to 8x memory can be used with `Row` compared to the `CompactRow`. However, if there are only a few rows that will be temporarily in the cache during an epoch, it'll be more efficient instead due to no ser/de needed.

Let's leave this a further optimization for now. 🥰